astropy/coordinates/attributes.py

```diff
@@ -132,12 +150,19 @@ def __get__(self, instance, frame_cls=None):
                 converted = True
 
             if converted:
-                setattr(instance, "_" + self.name, out)
+                # Use object.__setattr__() to be able to modify frozen dataclasses
+                object.__setattr__(instance, "_" + self.name, out)
 
         return out
 
     def __set__(self, instance, val):
-        raise AttributeError("Cannot set frame attribute")
+        if is_dataclass(instance):
```

Comment:
I added more comments to the code

Reply:
For reference, I had to solve this exact issue recently. Here's my solution that returns `self` except when a dataclass is setting the default (which in this case is not optional):

```python
def __get__(
        self: Self,
        instance: "T | None",
        owner: "type[T] | None",
    ) -> Self | Value:
        # Get from class
        if instance is None:
            # If this is being set as part of a dataclass constructor,
            # then we raise an AttributeError if there is no default value. This
            # is to prevent this instance from being set as the default value of
            # the dataclass field and erroneously included in the class'
            # ``__init__`` signature.
            if not is_dataclass(owner) or self.name not in owner.__dataclass_fields__:
                if self.default is Sentinel.MISSING:  # Sentinel.MISSING is an enum
                    raise AttributeError
                return self.default
            return self

        # Get from instance
        return instance.__dict__[self.name]

```

Reply:
https://github.com/GalacticDynamics/galax/blob/aa1c572bf749634e87f71b48b70136b43eb1a7ff/src/galax/potential/_potential/param/field.py#L120-L139